File: quicksort.ex
defmodule Quicksort do
  def sort([]), do: []
  def sort([pivot | rest]) do
    less    = Enum.filter(rest, &(&1 <= pivot))
    greater = Enum.filter(rest, &(&1 > pivot))
    sort(less) ++ [pivot] ++ sort(greater)
  end
end

IO.inspect(Quicksort.sort([3, 6, 8, 10, 1, 2, 1]))

File: quicksort.clj
(defn quicksort [coll]
  (if (< (count coll) 2)
    coll
    (let [pivot (first coll)
          rest  (rest coll)
          less  (filter #(<= % pivot) rest)
          greater (filter #(> % pivot) rest)]
      (concat (quicksort less) [pivot] (quicksort greater)))))

(println (quicksort [3 6 8 10 1 2 1]))
